### PR TITLE
update the adr with the new syntax

### LINKED ISSUE #1608: Extend ADR002 with the syntax for new types
Extend ADR002 with the new syntax as introduced in #1602.

### PR DESCRIPTION
**Pull request by konnov**
Closes #1608. This PR extends ARD002 with the new syntax for records, variants, and rows. It follows the syntax introduced in the PR #1601.

- [x] Documentation added for any new functionality

### CODE REVIEW

File: docs/src/adr/002adr-types.md [hunk 1]
```diff
@@ -127,8 +127,87 @@ This rule binds a type (produced by `T`) to a name (produced by `typeConst`). As
...
+    // a variant option with a partially defined structure
+    //   (a variant option over a row).
+    | { tag: stringLiteral, field: T2, ..., field: T2, typeVar }
+
+// Special syntax for the rows, which is internal to the type checker.
+row ::=
+    // A row with a fully defined structure.
+    | '(|' field ':' T2 ',' ...',' field ':' T2 '|)'
+    // A row with a partially defined structure (ending with a row).
+    | '(|' field ':' T2 ',' ...',' field ':' T2 ',' typeVar '|)'
```
Comment (gabrielamafra) [inline docs/src/adr/002adr-types.md:173]: I believe we are using pipes as separators for rows

```suggestion
    | '(|' field ':' T2 '|' ...'|' field ':' T2 '|)'
    // A row with a partially defined structure (ending with a row).
    | '(|' field ':' T2 '|' ...'|' field ':' T2 '|' typeVar '|)'
```

File: docs/src/adr/002adr-types.md [hunk 2]
```diff
@@ -127,8 +127,87 @@ This rule binds a type (produced by `T`) to a name (produced by `typeConst`). As
...
+**Examples.**
+
+* `r1` is a record that has the fields `a` and `b`, where `a` is an integer and
+  `b` is a string. Its type is `{ a: Int, b: Str }`.
+
+* `r2` is a record that has the fields `a` of type `Int` and `b` of type `Str`
+  and other fields, whose precise structure is captured with a type variable
+  `c`. The type of `r2` is `{ a: Int, b: Str, c }`.  More precisely, the
+  variable `c` should be a row. For instance, `c` can be equal to the row `(|
+  f: Bool, g: Set(Int) |)`; in this case, `r2` would be a record of type `{ a:
```
Comment (gabrielamafra) [inline docs/src/adr/002adr-types.md:185]: Also here
```suggestion
  variable `c` should be a row. For instance, `c` can be equal to the row `(|
  f: Bool | g: Set(Int) |)`; in this case, `r2` would be a record of type `{ a:
```

File: docs/src/adr/002adr-types.md [hunk 3]
```diff
@@ -127,8 +127,87 @@ This rule binds a type (produced by `T`) to a name (produced by `typeConst`). As
...
+     `tag` is equal to `"A"`).
+
+   - It has the fields `tag` of type `Str` and `b` of type `Bool` (if the field
+     `tag` is equal to `"B"`).
+
+* `v2` is an empty variant, which admits no options. It has the type
+  `Variant()`.
+
+* `v3` is a variant whose structure is defined by the type variable `b`.
+   The type of `v3` is `Variant(b)`. Note that `b` is a type variable should be
```
Comment (gabrielamafra) [inline docs/src/adr/002adr-types.md:200]: Did you mean
```suggestion
   The type of `v3` is `Variant(b)`. Note that `b` type variable should be
```

File: docs/src/adr/002adr-types.md [hunk 4]
```diff
@@ -127,8 +127,87 @@ This rule binds a type (produced by `T`) to a name (produced by `typeConst`). As
...
+```
+// Type System 1.2
+T2 ::=
+    // all types of Type System 1
+    T
+    // A new record type with a fully defined structure.
+    // The set of fields may be empty.
+    | '{' field ':' T2 ',' ...',' field ':' T2 '}'
+    // A new record type with a partially defined structure
+    // (the type variable should be a 'row').  // The set of fields may be empty.
```
Comment (gabrielamafra) [inline docs/src/adr/002adr-types.md:152]: We should break a line or remove the extra `//`
```suggestion
    // (the type variable should be a 'row').
    // The set of fields may be empty.
```

File: docs/src/adr/002adr-types.md [hunk 5]
```diff
@@ -127,8 +127,88 @@ This rule binds a type (produced by `T`) to a name (produced by `typeConst`). As
 of `typeConst`, the name should be an identifier in the upper case. The type checker should use the bound type instead
 of the constant type. For examples, see [Section 2.4](#useTypeAlias).
 
+<a id="rows"></a>
+### 1.3. Type System 1.2, including precise records, variants, and rows
```
Comment (thpani) [inline docs/src/adr/002adr-types.md:131]: For later use:

```suggestion
### <a id="ts-1.2"></a>1.3. Type System 1.2, including precise records, variants, and rows
```

File: docs/src/adr/002adr-types.md [hunk 6]
```diff
@@ -149,45 +229,55 @@ The parser only supports one-line comments that starts with `//`. Since type
...
 
 ```tla
-{[type |-> "1a", bal |-> 1],
- [type |-> "2a", bal |-> 2, val |-> 3]}
+{[tag |-> "1a", bal |-> 1],
+ [tag |-> "2a", bal |-> 2, val |-> 3]}
 ```
 
-would be something like:
+In Type System 1.2 (Section 1.3), this set has the type of a set over a variant
```
Comment (thpani) [inline docs/src/adr/002adr-types.md:256]:
```suggestion
In Type System 1.2 ([Section 1.3](#ts-1.2)), this set has the type of a set over a variant
```

File: docs/src/adr/002adr-types.md [hunk 7]
```diff
@@ -149,45 +229,55 @@ The parser only supports one-line comments that starts with `//`. Since type
...
+TLA+, such pattern is a set comprehension, e.g., `{ r \in S: r.tag = "1a" }`.
+In the typed version, we define a minimal set of operators over variants in the
+module
+[Variants.tla](https://github.com/informalsystems/apalache/blob/unstable/src/tla/Variants.tla).
+For instance, instead of writing the set comprehension, we have to use a filter
+over a set of variants: `FilterByTag(S, "1a")`.
+
+**Other type systems.**
+Type System 1 is also very much in line with the [type system by Stephan Merz and Hernan Vanzetto](https://dblp.org/search?q=Automatic+Verification+of+%7BTLA%7D+%2B+Proof+Obligations+with+%7BSMT%7D+Solvers)
+, which is used internally by
```
Comment (thpani) [inline docs/src/adr/002adr-types.md:274]:
```suggestion
Type System 1 is also very much in line with the [type system by Stephan Merz and Hernan Vanzetto](https://dblp.org/search?q=Automatic+Verification+of+%7BTLA%7D+%2B+Proof+Obligations+with+%7BSMT%7D+Solvers),
which is used internally by
```

File: docs/src/adr/002adr-types.md [hunk 8]
```diff
@@ -149,45 +229,55 @@ The parser only supports one-line comments that starts with `//`. Since type
...
+[TLAPS](https://tla.msr-inria.inria.fr/tlaps/content/Home.html) when translating proof obligations in SMT. We introduce
+types for user-defined operators, on top of their types for TLA+ expressions that do not contain user-defined operators.
+
+We expect that this type system will evolve in the future. That is why we call
+it __Type System 1__. Section 1.3 presents its extension to __Type System
```
Comment (thpani) [inline docs/src/adr/002adr-types.md:279]:
```suggestion
it __Type System 1__. [Section 1.3](#ts-1.2) presents its extension to __Type System
```

File: docs/src/adr/002adr-types.md [hunk 9]
```diff
@@ -149,45 +229,55 @@ The parser only supports one-line comments that starts with `//`. Since type
...
+1.2__. Feel free to suggest __Type System 2.0__ :-)
```
Comment (thpani) [inline docs/src/adr/002adr-types.md:280]: Was there a Type System 1.1?

**Review by thpani — APPROVED** [review]
👍 LGTM

File: docs/src/adr/002adr-types.md (hunk 9 above)
Comment (konnov) [reply docs/src/adr/002adr-types.md:280]: Nope, we are going directly from 1 to 1.2 :))

Comment (thpani) [reply docs/src/adr/002adr-types.md:280]: Why? Isn't this confusing?

Comment (konnov) [reply docs/src/adr/002adr-types.md:280]: May be. I don't find it really confusing though. The most important property of versions is monotonicity.

**Comment by konnov** [comment]
Merging this one

**Pull request merged**